src/Shared/Model/ArtifactUri.cs

```diff
@@ -0,0 +1,47 @@
+// Copyright (c) Microsoft Corporation. Licensed under the MIT License.
+
+namespace Microsoft.CST.OpenSource.Model;
+
+using System;
+using System.IO;
+
+public struct ArtifactUri
+{
+    public ArtifactUri(ArtifactType type, Uri uri, string? extension = null)
+    {
+        Type = type;
+        Uri = uri;
+        Extension = extension ?? Path.GetExtension(uri.AbsolutePath);
+    }
+
+    public ArtifactUri(ArtifactType type, string uri, string? extension = null)
+    {
+        Type = type;
+        Uri = new Uri(uri);
+        Extension = extension ?? Path.GetExtension(uri);
+    }
+    
+    public ArtifactType Type { get; }
+    public Uri Uri { get; }
+    public string Extension { get; }
+
+    public enum ArtifactType
+    {
+        Unknown = 0,
+        Tarball,
+        ZipFile,
+        Binary,
+
+        // NuGet specific
+        Nupkg,
+        Nuspec,
+
+        // PyPI specific
+        Wheel,
+
+        // Maven specific
+        Pom,
+        Jar,
+        Sources,
+    }
```

Comment:
I don't want to get too clever if it complicates things, but I wonder whether the `ArtifactUri` struct could be generic across the enum type? So instead of defining `ArtifactType` here, the `Type` field on this class would have a generic `enum` type?

That would let the NuGet-specific classes return NuGet specific artifact type values, Maven-specific classes could return Maven-specific values, etc.

Reply:
I hadn't thought about that, I like it.

Reply:
This is closer, but as written right now, the compiler doesn't offer users of this field any assistance in terms of what the value might be.

This is more along the lines of what I was getting at:

```c#
public readonly record struct ArtifactUri<T> where T : System.Enum
{
    // <snip />
    T Type {get;}
   
   // <snip />
}
```

That way, the compiler can offer assistance to users of this class to know what possible Enum values could be returned. The tradeoff being that we would have to propagate the generic parameter through appropriate places.

What do you think?

Reply:
The revised `ArtifactUri` approach looks great - thanks for making the changes! Resolving this thread.